synapse/util/caches/response_cache.py

```diff
@@ -80,6 +81,42 @@ def get(self, key: T) -> Optional[defer.Deferred]:
             self._metrics.inc_misses()
             return None
 
+    @staticmethod
+    def _safe_call(func: Callable[[Any], bool], ret_val: Any, key: T) -> bool:
+        """
+        Wraps a conditional checking function with a try-catch, returns False
+        if the conditional function threw an exception during evaluation.
+        """
+        try:
+            return func(ret_val)
+        except Exception as e:
+            logger.warning(
+                "conditional %r threw exception on %s, ignoring and invalidating cache...",
+                func,
+                key,
+                exc_info=e,
+            )
+            return False
+
+    def _remove_cache(self, key: T):
+        self.pending_result_cache.pop(key, None)
+
+    def _errback_remove_cache(self, f: Failure, key: T):
+        self._remove_cache(key)
+        return f
+
+    def _test_cache(self, r: Any, key: T):
+        should_cache = all(
+            self._safe_call(func, r, key=key)
+            for func in self.pending_conditionals.pop(key, [])
+        )
```

Comment:
for links: I've argued in #synapse-dev and https://github.com/matrix-org/synapse/pull/9358/files#r585481702 that we should only support one conditional per key, which simplifies this code.

Reply:
I'll add that to the "small changes to `ResponseCache` PR", this one is focused on fixing #9507, i dont wanna drag everything else into here.

Reply:
Created #9525 for this